include/swift/AST/Types.h

```diff
@@ -2899,17 +2901,28 @@ class AnyFunctionType : public TypeBase {
     ExtInfo(Representation Rep,
             bool IsNoEscape,
             // SWIFT_ENABLE_TENSORFLOW
-            bool Throws, bool IsDifferentiable)
+            bool Throws, DifferentiabilityKind diffKind)
       : ExtInfo(Rep, Throws) {
       Bits |= (IsNoEscape ? NoEscapeMask : 0);
       // SWIFT_ENABLE_TENSORFLOW
-      Bits |= (IsDifferentiable ? DifferentiableMask : 0);
+      Bits |=
+          ((unsigned)diffKind << DifferentiableMaskOffset) & DifferentiableMask;
     }
 
     bool isNoEscape() const { return Bits & NoEscapeMask; }
     bool throws() const { return Bits & ThrowsMask; }
     // SWIFT_ENABLE_TENSORFLOW
-    bool isDifferentiable() const { return Bits & DifferentiableMask; }
+    bool isDifferentiable() const {
+      return ((DifferentiabilityKind)DifferentiableMask ==
+              DifferentiabilityKind::Normal) || isLinear();
+    }
+    bool isLinear() const {
+      return ((DifferentiabilityKind)DifferentiableMask ==
+              DifferentiabilityKind::Linear);
+    }
+    DifferentiabilityKind getDifferentiableMask() const {
```

Comment:
Hmm, this is incorrect. A mask has no meaning. It's just some set bits in a random spot. It should never be coerced into a `DifferentiabilityKind`. Instead, we should define a `getDifferentiabilityKind()` and return `DifferentiabilityMask(DifferentiableMask << DifferentiableOffset)`.

Reply:
Ya this seems like the one bug I had, but now I have another one that I'm looking into. And just making sure shouldn't it be `DifferentiableMask >> DifferentiableOffset` since we want to undo the offset?

Reply:
Sorry, please disregard my earlier suggestion.

Also, I'd suggest doing the following:
* Rename `DifferentiableMask` to `DifferentiabilityMask` and `DifferentiableMaskOffset` to `DifferentiabilityMaskOffset`, since this represents something more than just a binary flag.
* Add a `NumDifferentiabilityMaskBits = 3` right below `DifferentiableMask` in the enum.
* In `getDifferentiabilityKind()`, return `DifferentiabilityKind((Bits & DifferentiableMask) >> DifferentiableMaskOffset)`.

Reply:
Shouldn't `NumDifferentiabilityMaskBits = 2` since `3=0x3=0b11` so it is 2 bits?

Reply:
Ah yes, I had a typo.

Reply:
What do we need `NumDifferentiabilityMaskBits` for? For `NumMaskBits` I see it is used for the following assert:
```c++
static_assert(ExtInfo::NumMaskBits == NumAFTExtInfoBits,
              "ExtInfo and AnyFunctionTypeBitfields must agree on bit size");
```
Should something like this be done for `NumDifferentiabilityMaskBits` as well?

Reply:
Something asserting that the mask, offset, and num-bits are consistent would be good, but I'm not sure if something like that can be statically evaluable.